Implement support for do..while

SamuelYvon opened the pull request:
I had heard about pnut from Marc. Too much fun not to participate. Here's a little contribution; I did not take too much time investigating the ins and outs of the codebase so I might be doing something wrong; feel free to correct me.

There's one limitation I am still trying to fix; The global `set -e` does not apply to the code output within the while.

As an aside, I noticed that `void main(void)` is not a valid signature for `pnut`, but it is one of the two valid signatures for main (ref https://en.cppreference.com/w/c/language/main_function)

laurenthuberdeau commented on README.md:
```diff
@@ -116,7 +116,6 @@ Unfortunately, certains C constructs don't map nicely to POSIX shell which means
 - The preprocessor is not perfect and may fail on some edge cases. `#if` and `#elif` are not supported. `#include <...>` are ignored.
 - All local variable declarations must be at the beginning of a function.
 - Aggregate types (arrays and structures) cannot be stack-allocated, passed by value or nested in a structure.
-- `do { ... } while(...)` is not supported at the moment.
```
🎉

laurenthuberdeau commented:
> As an aside, I noticed that void main(void) is not a valid signature for pnut, but it is one of the two valid signatures for main (ref https://en.cppreference.com/w/c/language/main_function)

I have a branch somewhere that supports these declarations, mainly to be able to compile TCC. There are a lot of small things that are wrong with the grammar pnut is accepting, this being a good example :)

SamuelYvon commented on sh.c:
```diff
@@ -1897,6 +1897,22 @@ void comp_statement(ast node, int else_if) {
+
+    nest_level -= 1;
+    loop_nesting_level -= 1;
+    append_glo_decl(wrap_str_lit("do :; done"));
```
> creating a temporary variable that would force the first iteration of the loop

That was my first idea; but I wasn't sure if the usage of a temporary was acceptable. It makes the readability a bit harder.

> If I'm reading the code correctly, your solution generates something like this: ...

This is exactly correct; 

>  In that same thread, another solution is presented which I find...

I'll try that out; it might also solve the `set -e` limitation

laurenthuberdeau replied:
> That was my first idea; but I wasn't sure if the usage of a temporary was acceptable. It makes the readability a bit harder.

Temporaries are used for nested function calls since shells don't have something equivalent. For example, this 1-line C function:

```C
#define wrap_char(c) (-c)

text global_var(ast ident_tok) {
  return string_concat(wrap_char('_'), wrap_str_pool(get_val(ident_tok)));
}
```

is compiled to:

```shell
: $((__t1 = ident_tok = 0))
_global_var() { let ident_tok $2
  let __t1 # !Notice that the temporary __t1 is treated as a local var!
  _get_val __t1 $ident_tok
  _wrap_str_pool __t1 $__t1
  _string_concat $1 $((-(__UNDERSCORE__))) $__t1
  endlet $1 __t1 ident_tok
}
```

They can be generated with the `fresh_ident` function but they are not free in terms of performance and, as you note, generally reduce readability.

SamuelYvon commented on examples/non_zero.c:
```diff
@@ -0,0 +1,13 @@
+  } while( n == '0' || !(n >= '0' && n <= '9'));
+
+  printf("You entered %c: bye bye!\r\n", n);
+}
```
Ok! I'll move to a test and / or keep this as an example and create a testable test program.

SamuelYvon commented:
@laurenthuberdeau I
- Added a simple test with a `do...while`
- Created an example (can remove if you don't care for a do while example)
- Changed the generated code of the `do..while` to match your preferred form, which also solves the issues with the `set -e` (win win)

LMK if there's anything else!

laurenthuberdeau commented on sh.c:
```diff
@@ -1898,7 +1898,7 @@ void comp_statement(ast node, int else_if) {
     append_glo_decl(wrap_str_lit("done"));
   } else if (op == DO_KW) {
-    append_glo_decl(wrap_str_lit("while"));
+    append_glo_decl(wrap_str_lit("while true; do"));
```
Very small nit: Using `for(;;;)` generates `while :; do` instead of `while true; do` (`:` and `true` are both equivalent). Do you mind changing the [`for` case](https://github.com/udem-dlteam/pnut/blob/master/sh.c#L1907) to use `true` for consistency?

laurenthuberdeau replied:
Good point, `true` isn't necessarily a built-in. Let's use `:` then.

SamuelYvon replied:
Done! :)

laurenthuberdeau reviewed (approved):
Very nice!

laurenthuberdeau commented:
If you're looking for more, the `exe` backend could also get `do...while();` [support](https://github.com/udem-dlteam/pnut/blob/master/exe.c#L1776-L1827).

It would also make it possible to remove the [temporary variable](https://github.com/udem-dlteam/pnut/blob/master/pnut.c#L1645-L1646) used in `get_tok` to simulate a do while loop which would certainly improve performance because `get_tok` would become a function with no local vars (and so it would avoid the overhead of simulating it).

SamuelYvon commented:
You can expect the PR for the exe backend over the weekend ;) If I forget shoot me an email :)

SamuelYvon commented:
Also, do you want me to add the changes for the exe on top of this or you'd rather another PR (and I can't merge btw, so that's on your schedule ;) )

laurenthuberdeau reviewed (changes requested):
I just realized that `DO_KW` needs to be handled in `mark_mutable_variables_statement`. This function is `ifdefed` away by default in pnut, but when the `OPTIMIZE_CONSTANT_PARAM` macro is defined, it traverses the AST to mark variables as mutable/immutable.

You can run the test with this option: `PNUT_OPTIONS="-DOPTIMIZE_CONSTANT_PARAM" ./run-tests.sh sh` to see the failure.

I've forgot a few times in the past, I'll add a CI check.

SamuelYvon commented:
Ok! Ping me when you've added the check; I'll check that when I get a second :)

laurenthuberdeau commented on sh.c:
```diff
@@ -2043,6 +2043,9 @@ void mark_mutable_variables_statement(ast node) {
     if (get_child(node, 1)) mark_mutable_variables_body(get_child(node, 1));
+  } else if (op == DO_KW) {
+    mark_mutable_variables_statement(get_child(node, 0));
+    if (get_child(node, 1)) mark_mutable_variables_body(get_child(node, 1));
```
I think in this case the child 0 is optional and child 1 is non-optional, so the opposite of the `WHILE_KW` case above. See https://github.com/udem-dlteam/pnut/blob/master/pnut.c#L3020-L3034

SamuelYvon replied:
Yep, wasn't sure of the structure, meant to ask you. Working on the `exe` backend right now.

SamuelYvon commented:
@laurenthuberdeau Added the `exe` backend. Did not change the main loop because the while has a second variable (macro something) and I didnt have time to check how to deal with it. 

LMK your toughts!

laurenthuberdeau commented on exe.c:
```diff
@@ -1824,6 +1824,32 @@ void codegen_statement(ast node) {
+    def_label(lbl1);
+
+    if (get_child(node, 0)) {
+      codegen_statement(get_child(node, 0));
```
`codegen_statement` returns early if it's argument is null :)

SamuelYvon replied:
Cool!

laurenthuberdeau commented on tests/_exe/do_while.c:
```diff
@@ -12,6 +12,18 @@ int bad_factorial(int num) {
+    putchar('O');
+  } else {
+    putchar('B');
+  }
```
Since `tests/_exe/do_while.c` and `tests/_sh/do_while.c` are the same modulo `printf`, I think we can share the same test file for both backend and move it to `tests/_all/do_while.c`

SamuelYvon replied:
Oh! Was not aware!

laurenthuberdeau commented on tests/_sh/do_while.golden:
```diff
@@ -0,0 +1,2 @@
+Factorial of 5: 120
+Factorial of 10: 3628800
```
Leftover file

SamuelYvon replied:
Sorry! 🤦🏼
